[Structural] Adding initial stress and strain capability

AlejandroCornejo opened the pull request:
In this PR I'll be adding to all the implemented CL the capability of imposing an initial strain or stress. Only the Linear elastic 3D law has been improved so far so you can see how it works.

In order to apply this initial strain/stress we have used the mdpa feature of 

Begin ElementalData INITIAL_STRESS_VECTOR
1  [6] (0,0,1e6,0,0,0)
2  [6] (0,0,1e6,0,0,0)
3  [6] (0,0,1e6,0,0,0)
4  [6] (0,0,1e6,0,0,0)
End ElementalData

Begin ElementalData INITIAL_STRAIN_VECTOR
1  [6] (0.01,0.01,0.01,0,0,0)
2  [6] (0.01,0.01,0.01,0,0,0)
3  [6] (0.01,0.01,0.01,0,0,0)
4  [6] (0.01,0.01,0.01,0,0,0)
End ElementalData

or by using the python process inside the json:
~~~python
{
{
            "python_module" : "set_initial_state_process",
            "kratos_module" : "KratosMultiphysics",
            "process_name"  : "set_initial_state_process",
            "Parameters"    : {
                    "mesh_id"         : 0,
                    "model_part_name" : "Structure",
                    "dimension"       : 2,
                    "imposed_strain"  : [0.0,0.00,0],
                    "imposed_stress"  : [1000000,0,0],
                    "imposed_deformation_gradient"  : [[1,0],[0,1]],
                    "interval"        : [0.0, 1e30]
                    }
        }
~~~

The method (inside linear elastic CL) checks whether the geometry has this initial stress/strain, otherwise is a ZeroVector:
~~~c++
    /**
     * @brief Adds the initial stress vector if it is defined in the InitialState
     */
    const void AddInitialStressVectorContribution(Vector& rStressVector, Parameters& rParameterValues) 
    {
        const auto p_initial_state = pGetpInitialState();
        if (p_initial_state) {
            noalias(rStressVector) += p_initial_state->GetInitialStressVector();
        }
    }
~~~

Additionally I can add the capability of retrieving the initial strains and stresses from the mat props.

AlejandroCornejo committed:
adding method to base constitutive class and one variable in the core

AlejandroCornejo committed:
changing the var

AlejandroCornejo committed:
another var

AlejandroCornejo committed:
removing

AlejandroCornejo committed:
1st working version

AlejandroCornejo committed:
optimizing

philbucher commented on applications/StructuralMechanicsApplication/custom_constitutive/elastic_isotropic_3d.h:
```diff
@@ -151,6 +152,32 @@ class KRATOS_API(STRUCTURAL_MECHANICS_APPLICATION) ElasticIsotropic3D
+    const Vector GetInitialStrainVector(Parameters& rParameterValues)
+    {
+        const auto &r_geometry = rParameterValues.GetElementGeometry();
+        if (r_geometry.Has(INITIAL_STRAIN_VECTOR)) {
```
hm so this is an additional `if` in each access

can you please make some performance tests as to how much this influences

AlejandroCornejo replied:
How do you do this test?

loumalouomega replied:
That's why in my branch I created an entire class with the imposed deformations. The idea would be to add to the current CL a template , so derive by default from ConstitutiveLaw, and being possible to derive from ConstitutiveLawWithImposedDeformation. Then there will not be overhead

AlejandroCornejo replied:
This is a philosofical question, it is better to generalize the existing Linear elastic Law with this initial strains or stresses or to create a derived one with this new capability. In my opinion I think that the 1st option is the less confusing (also less quantity of basic CL) but the overhead is necessary right?

loumalouomega replied:
> This is a philosofical question, it is better to generalize the existing Linear elastic Law with this initial strains or stresses or to create a derived one with this new capability. In my opinion I think that the 1st option is the less confusing (also less quantity of basic CL) but the overhead is necessary right?

It is necessary unless you separate the laws between the ones with initial strain and the ones without it as I said, which has it disadvantages

AlejandroCornejo replied:
What's your opinion @RiccardoRossi ?

rubenzorrilla replied:
Even though I have a limited knowledge about the CL implementation, I support @philbucher 's and @loumalouomega 's observations as this implies introducing an if statement per Gauss pt.

philbucher commented on applications/StructuralMechanicsApplication/custom_constitutive/elastic_isotropic_3d.cpp:
```diff
@@ -60,17 +60,21 @@ void  ElasticIsotropic3D::CalculateMaterialResponsePK2(ConstitutiveLaw::Paramete
     }
 
+    const Vector& r_initial_strain = this->GetInitialStrainVector(rValues);
+    r_strain_vector -= r_initial_strain;
```
it would be more efficient to check here if initial stress/strain was specified

AlejandroCornejo replied:
mm Why? the initial strain or stress can be imposed separatedly

philbucher reviewed (changes requested):
comments regarding performance, otherwise looks good to me

blocking to prevent accidental merge as this affects many

philbucher commented on applications/StructuralMechanicsApplication/custom_constitutive/elastic_isotropic_3d.h:
```diff
@@ -151,6 +152,32 @@ class KRATOS_API(STRUCTURAL_MECHANICS_APPLICATION) ElasticIsotropic3D
+    /**
+     * @brief Returns the initial stress vector if it is defined in the geometry
+     */
+    const Vector GetInitialStressVector(Parameters& rParameterValues)
```
```suggestion
    Vector GetInitialStressVector(Parameters& rParameterValues) const
```

philbucher commented on applications/StructuralMechanicsApplication/custom_constitutive/elastic_isotropic_3d.h:
```diff
@@ -151,6 +152,32 @@ class KRATOS_API(STRUCTURAL_MECHANICS_APPLICATION) ElasticIsotropic3D
+    /**
+     * @brief Returns the initial strain vector if it is defined in the geometry
+     */
+    const Vector GetInitialStrainVector(Parameters& rParameterValues)
```
```suggestion
    Vector GetInitialStrainVector(Parameters& rParameterValues) const
```

AlejandroCornejo replied:
I don't know why if I change this I get a compilation error

loumalouomega reviewed (changes requested):
Just the comments from @philbucher ... and I don't see a test

AlejandroCornejo committed:
moving the const

AlejandroCornejo committed:
not const

AlejandroCornejo committed:
Merge branch 'master' into Structural/adding-initial-stress-strain-vectors

AlejandroCornejo committed:
optimizing the method

AlejandroCornejo committed:
Merge branch 'master' into Structural/adding-initial-stress-strain-vectors

AlejandroCornejo committed:
Merge branch 'master' into Structural/adding-initial-stress-strain-vectors

AlejandroCornejo committed:
Merge branch 'master' into Structural/adding-initial-stress-strain-vectors

AlejandroCornejo committed:
Merge branch 'master' into Structural/adding-initial-stress-strain-vectors

AlejandroCornejo committed:
CREATING CLASS -> initial state

AlejandroCornejo committed:
changing place

AlejandroCornejo committed:
adding methods

AlejandroCornejo committed:
more methods

AlejandroCornejo committed:
adding nullptr to constitutitve_law.h

AlejandroCornejo committed:
minor

AlejandroCornejo committed:
modify method

AlejandroCornejo committed:
good way of doing it

AlejandroCornejo committed:
creating process

AlejandroCornejo committed:
updated version and process

AlejandroCornejo committed:
adding variable

AlejandroCornejo committed:
correcting

AlejandroCornejo committed:
removing unused var

AlejandroCornejo committed:
compiling version

AlejandroCornejo committed:
removing const

AlejandroCornejo committed:
adding const

AlejandroCornejo committed:
template to process

AlejandroCornejo committed:
add variable to .h

AlejandroCornejo committed:
add process to python

AlejandroCornejo committed:
detailing

philbucher commented on applications/StructuralMechanicsApplication/custom_constitutive/elastic_isotropic_3d.h:
```diff
@@ -19,6 +19,7 @@
 
 // Project includes
 #include "includes/constitutive_law.h"
+#include "includes/mat_variables.h"
```
```suggestion
```

unused

AlejandroCornejo replied:
done

philbucher commented on kratos/includes/initial_state.h:
```diff
@@ -0,0 +1,249 @@
+        enum class InitialImposingType {StrainOnly = 0, StressOnly = 1, DeformationGradientOnly = 2, StrainAndStress = 3, DeformationGradientAndStress = 4};
+
+        /// Pointer definition of NodeSearchUtility
+        KRATOS_CLASS_POINTER_DEFINITION(InitialState);
```
shouldn't this be an intrusive pointer @RiccardoRossi ?

philbucher commented on applications/StructuralMechanicsApplication/custom_constitutive/elastic_isotropic_3d.h:
```diff
@@ -151,6 +152,28 @@ class KRATOS_API(STRUCTURAL_MECHANICS_APPLICATION) ElasticIsotropic3D
+     */
+    const void AddInitialStressVectorContribution(Vector& rStressVector, Parameters& rParameterValues) 
+    {
+        const auto p_initial_state = this->GetpInitialState();
```
```suggestion
        const auto p_initial_state = this->pGetInitialState();
```

AlejandroCornejo replied:
ok

philbucher commented on applications/StructuralMechanicsApplication/custom_constitutive/elastic_isotropic_3d.h:
```diff
@@ -151,6 +152,28 @@ class KRATOS_API(STRUCTURAL_MECHANICS_APPLICATION) ElasticIsotropic3D
+    const void AddInitialStressVectorContribution(Vector& rStressVector, Parameters& rParameterValues) 
+    {
+        const auto p_initial_state = this->GetpInitialState();
+        if (p_initial_state != nullptr) {
```
```suggestion
        if (p_initial_state) {
```

AlejandroCornejo replied:
fancy

philbucher commented on applications/StructuralMechanicsApplication/custom_constitutive/elastic_isotropic_3d.h:
```diff
@@ -312,6 +335,7 @@ class KRATOS_API(STRUCTURAL_MECHANICS_APPLICATION) ElasticIsotropic3D
 
     ///@name Protected static Member Variables
     ///@{
+    // static const Vector v6( ZeroVector(6) );
```
```suggestion
```

leftover

philbucher commented on kratos/includes/constitutive_law.h:
```diff
@@ -539,6 +542,22 @@ class KRATOS_API(KRATOS_CORE) ConstitutiveLaw : public Flags
+    /**
+     * @return The initial state of strains/stresses/F
+     */
+    InitialState* GetpInitialState()
```
```suggestion
    InitialState* pGetInitialState()
```

needs also a `const` version

philbucher commented on kratos/includes/initial_state.h:
```diff
@@ -0,0 +1,249 @@
+ * @details Storages the information regarding initial stresses/strains/F
+ * @author Alejandro Cornejo
+ */
+class KRATOS_API(KRATOS_CORE) InitialState
```
should be separaten into header/source

AlejandroCornejo replied:
okkey, i did not do it because in this folder almost all of them were only 1 file

philbucher replied:
check kratos/sources ;)

philbucher replied:
ping

philbucher commented on kratos/processes/set_initial_state_process.h:
```diff
@@ -0,0 +1,190 @@
+ *  
+*/
+template<SizeType TDim>
+class SetInitialStateProcess : public Process
```
header/source

AlejandroCornejo replied:
IDEM

philbucher replied:
except a few leftovers all `Process`es are splitted

philbucher commented on kratos/processes/set_initial_state_process.h:
```diff
@@ -0,0 +1,190 @@
+        const auto it_elem_begin = mrModelPart.ElementsBegin();
+        const auto& integration_points = it_elem_begin->GetGeometry().IntegrationPoints(it_elem_begin->GetIntegrationMethod());
+
+        #pragma omp parallel for
```
please use parallel utilities

AlejandroCornejo replied:
do you have an example? :)

philbucher replied:
Search for block_for_each, there are quite a few usages now

philbucher commented on applications/StructuralMechanicsApplication/custom_constitutive/elastic_isotropic_3d.h:
```diff
@@ -151,6 +152,28 @@ class KRATOS_API(STRUCTURAL_MECHANICS_APPLICATION) ElasticIsotropic3D
+    /**
+     * @brief Adds the initial stress vector if it is defined in the InitialState
+     */
+    const void AddInitialStressVectorContribution(Vector& rStressVector, Parameters& rParameterValues) 
```
cannot those functions be `static` functions of the `ConstitutiveLaw`?
then they don't have to be reimplemented in every element and this would probably also solve the downcasting issue

AlejandroCornejo replied:
I've tried and I've noticed that inside this method there's a return of a non static `mpInitialState `so they cannot be static I think

philbucher replied:
Ok then without the static would be working I think

philbucher reviewed (commented):
some more comments

philbucher commented on kratos/includes/initial_state.h:
```diff
@@ -0,0 +1,249 @@
+        /**
+         * @brief This method returns the initial strain vector if was set before
+         */
+        const Vector& GetInitialStrainVector() {
```
```suggestion
        const Vector& GetInitialStrainVector() const 
        {
```

philbucher commented on kratos/includes/initial_state.h:
```diff
@@ -0,0 +1,249 @@
+        /**
+         * @brief This method returns the initial stress vector if was set before
+         */
+        const Vector& GetInitialStressVector() {
```
```suggestion
        const Vector& GetInitialStressVector() const 
        {
```

philbucher commented on kratos/includes/initial_state.h:
```diff
@@ -0,0 +1,249 @@
+        /**
+         * @brief This method returns the initial stress vector if was set before
+         */
+        const Matrix& GetInitialDeformationGradientMatrix() {
```
```suggestion
        const Matrix& GetInitialDeformationGradientMatrix() const
        {
```

philbucher reviewed (commented):
missing `const`

AlejandroCornejo committed:
Merge branch 'master' into Structural/adding-initial-stress-strain-vectors

AlejandroCornejo committed:
accessing to the CL by means of CalculateOnIntegrationPoints :)

AlejandroCornejo committed:
unused include

philbucher commented on kratos/processes/set_initial_state_process.h:
```diff
@@ -0,0 +1,191 @@
+            // Assign the values to the GP of the element
+
+            std::vector<ConstitutiveLaw::Pointer> constitutive_law_vector;
+            it_elem->CalculateOnIntegrationPoints(CONSTITUTIVE_LAW, constitutive_law_vector, mrModelPart.GetProcessInfo());
```
I am not sure if this flexible enough. This relies on a spcefic way the constitutive laws are implemented in an element. Note that this is NOT enforced through the baseclass of `Element`, hence I think we should also avoid this implicit restriction here
why don't you use `SetValuesOnIntegrationPoints` instead and let the element handle this?

AlejandroCornejo replied:
Hi @philbucher ! In fact the method  `CalculateOnIntegrationPoints` can be found in the `element.h` :D

philbucher replied:
but CONSTITUTIVE_LAW and especially the corresponding vectors are not!
Note that not all elements implement the Constitutive laws as a vector (e.g. the truss & the shells)

=> hence this is an implementation detail of the corresponding elements and should not be enforced

AlejandroCornejo replied:
The only thing that I dont really like is the fact of involving the elements in terms of Strains and Stresses... With the SetValuesOnIntegrationPoints I'll have to implement the interaction of the Element and the CL. 
I suppose that I'll have also to add a new variable like `INITIAL_STATE` in order to set everything.

AlejandroCornejo replied:
I think that @RiccardoRossi  has something to say about this! 👍

philbucher replied:
I spoke with @RiccardoRossi and we agreed to move forward with your current proposal
For the solids the `CONSTITUTIVE_LAW` is straight forward to use, and for the other elements we can internally construct it

FYI @KlausBSautter also the membrane has a `std::vector<ConstitutiveLaw::Pointer> mConstitutiveLawVector` too so it will also be straight forward for you 

Overall I think it makes sense to use this for future elements too if possible/applicable

AlejandroCornejo committed:
changes in elastic

AlejandroCornejo committed:
adding p_

AlejandroCornejo committed:
now static

AlejandroCornejo committed:
not static, no compiling

AlejandroCornejo committed:
adding const

AlejandroCornejo committed:
initial strains are substracted

AlejandroCornejo committed:
Merge branch 'master' into Structural/adding-initial-stress-strain-vectors

AlejandroCornejo committed:
Merge branch 'master' into Structural/adding-initial-stress-strain-vectors

AlejandroCornejo committed:
IDE tabs

AlejandroCornejo committed:
minor style

AlejandroCornejo committed:
adding more constructors

AlejandroCornejo committed:
NOW CORRECT

AlejandroCornejo committed:
cannot be a raw pointer

AlejandroCornejo committed:
silly error

AlejandroCornejo committed:
minor errors

AlejandroCornejo committed:
adding python process

AlejandroCornejo committed:
defaults

AlejandroCornejo committed:
implementing process python

AlejandroCornejo committed:
minor corrections

AlejandroCornejo committed:
for printing purposes

AlejandroCornejo committed:
reverting commit

AlejandroCornejo committed:
optimized and intrusive pointer

AlejandroCornejo committed:
corrected

AlejandroCornejo committed:
minor

AlejandroCornejo committed:
creating cpp process file

AlejandroCornejo committed:
process separate in .h and .cpp

AlejandroCornejo committed:
compiling

AlejandroCornejo committed:
last details

AlejandroCornejo committed:
adding override and test

AlejandroCornejo committed:
not override

AlejandroCornejo committed:
advancing test

AlejandroCornejo committed:
adding 3d test

AlejandroCornejo committed:
minor

AlejandroCornejo committed:
provisional test -> to be changed

AlejandroCornejo committed:
adding virtual destructor

AlejandroCornejo committed:
ading dummy CL

AlejandroCornejo committed:
minor compiling dummy cl

AlejandroCornejo committed:
adding new capabiliies for the test

AlejandroCornejo committed:
test provisional

AlejandroCornejo committed:
Merge branch 'master' into Structural/adding-initial-stress-strain-vectors

AlejandroCornejo committed:
improving test and rpocess

AlejandroCornejo committed:
correcting process

AlejandroCornejo committed:
adding initialize method to the test element

AlejandroCornejo committed:
solving 2D test

AlejandroCornejo committed:
whitespaces and adding 3D test

AlejandroCornejo committed:
solving one comp warning in pr

AlejandroCornejo committed:
more checks

AlejandroCornejo committed:
adding delete to copy

AlejandroCornejo committed:
removing leftover

AlejandroCornejo committed:
unused code

AlejandroCornejo committed:
setting properties to elem

AlejandroCornejo committed:
trying to passthe debug check

AlejandroCornejo committed:
clean version

AlejandroCornejo committed:
Merge branch 'master' into Structural/adding-initial-stress-strain-vectors

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_constitutive/elastic_isotropic_3d.h:
```diff
@@ -151,6 +151,28 @@ class KRATOS_API(STRUCTURAL_MECHANICS_APPLICATION) ElasticIsotropic3D
+     */
+    void AddInitialStressVectorContribution(Vector& rStressVector, Parameters& rParameterValues) 
+    {
+        const auto p_initial_state = pGetpInitialState();
```
this is inefficient because you increment and decrement the counter of the intrusive_ptr.

it would be better if we define in the base class an inline function "HasInitialState" and one "GetInitialState" returning a reference (by the way, is the second p willingly added?? in pGet**p**InitialState)

AlejandroCornejo replied:
addressed

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_constitutive/elastic_isotropic_3d.h:
```diff
@@ -151,6 +151,28 @@ class KRATOS_API(STRUCTURAL_MECHANICS_APPLICATION) ElasticIsotropic3D
+     */
+    void AddInitialStrainVectorContribution(Vector& rStrainVector, Parameters& rParameterValues)
+    {
+        const auto p_initial_state = pGetpInitialState();
```
same

AlejandroCornejo replied:
done

RiccardoRossi commented on kratos/includes/constitutive_law.h:
```diff
@@ -539,6 +542,22 @@ class KRATOS_API(KRATOS_CORE) ConstitutiveLaw : public Flags
+    /**
+     * @return The initial state of strains/stresses/F
+     */
+    void SetpInitialState(InitialState::Pointer pInitialState)
```
I would call this SetInitialState

AlejandroCornejo replied:
oído

RiccardoRossi commented on kratos/includes/initial_state.h:
```diff
@@ -0,0 +1,279 @@
+
+        // Selective constructor for vectors
+        InitialState(const Vector& rImposingEntity,
+                     const int InitialImposingType = 0)
```
you shall use an enum here. the usage will be very similar, but clearer

AlejandroCornejo replied:
changed

AlejandroCornejo committed:
adding props to elem

AlejandroCornejo committed:
Riccardo's requests

philbucher commented on kratos/includes/constitutive_law.h:
```diff
@@ -539,6 +542,41 @@ class KRATOS_API(KRATOS_CORE) ConstitutiveLaw : public Flags
+            return true;
+        else
+            return false;
+    }
```
```suggestion
    bool HasInitialState() const
    {
        return mpInitialState != nullptr;
    }
```

philbucher commented on kratos/python_scripts/assing_initial_state_process.py:
```diff
@@ -0,0 +1,89 @@
+        Keyword arguments:
+        self -- It signifies an instance of a class.
+        """
+        self.ExecuteInitializeSolutionStep()
```
why?
This should not be here IMO

AlejandroCornejo replied:
Should not be here the `ExecuteSolutionStep()`? and the KratosCore?

philbucher replied:
I mean ExecuteBeforeSolutionStep calls `ExecuteInitializeSolutionStep`?
I know this is done in some places for some weird reasons but I wouldn't add it to new code

AlejandroCornejo replied:
That's definetly true!! it was a mistake :) THX @philbucher

philbucher commented on applications/StructuralMechanicsApplication/custom_constitutive/elastic_isotropic_3d.h:
```diff
@@ -151,6 +151,28 @@ class KRATOS_API(STRUCTURAL_MECHANICS_APPLICATION) ElasticIsotropic3D
+            const auto& r_initial_state = GetInitialState();
+            noalias(rStrainVector) += r_initial_state.GetInitialStrainVector();
+        }
+    }
```
ping can you please try again to move this to the constitutive law?
Otherwise this will be duplicated in every law

AlejandroCornejo replied:
I can do it! @RiccardoRossi opinion?

philbucher reviewed (commented):
@KlausBSautter can you please have a look if this is compatible with the other elements?

AlejandroCornejo committed:
Update kratos/includes/constitutive_law.h

Co-authored-by: Philipp Bucher <philipp.bucher@tum.de>

AlejandroCornejo committed:
minor mistake in python process

philbucher commented on kratos/python_scripts/assing_initial_state_process.py:
```diff
@@ -0,0 +1,80 @@
+# Importing the Kratos Library
```
BTW the file name has a typo :)

AlejandroCornejo replied:
shit... thank you again hehe

AlejandroCornejo committed:
typo in process

AlejandroCornejo committed:
- is correct in strain

AlejandroCornejo committed:
adding the methods inside the constitutive_law

marandra committed:
Python process now accepts strings (e.g. '0.1*t') as vector components (only 't' for now)

marandra committed:
Merge branch 'Structural/adding-initial-stress-strain-vectors' of github.com:KratosMultiphysics/Kratos into Structural/adding-initial-stress-strain-vectors

marandra committed:
Black

marandra committed:
Fix SetInitialState is InInterval

marandra committed:
Codacy

philbucher commented on kratos/includes/initial_state.h:
```diff
@@ -0,0 +1,279 @@
+
+// System includes
+#include <iostream>
+#include <string>
```
```suggestion
#include <iostream>
#include <string>
#include <atomic>

```

necessary for the intrusive ptr to work

philbucher commented on kratos/processes/set_initial_state_process.cpp:
```diff
@@ -0,0 +1,177 @@
+            (aux_initial_strain, aux_initial_stress, aux_initial_F);
+
+        #pragma omp parallel for
+        for (int i = 0; i < static_cast<int>(mrModelPart.Elements().size()); i++) {
```
Parallel Utils

loumalouomega replied:
IDem

philbucher commented on kratos/processes/set_initial_state_process.h:
```diff
@@ -0,0 +1,153 @@
+#include "includes/element.h"
+#include "includes/model_part.h"
+#include "includes/initial_state.h"
+#include "includes/mat_variables.h"
```
move to cpp

philbucher replied:
```suggestion
#include "processes/process.h"
#include "includes/model_part.h"
#include "includes/mat_variables.h"
```

can be remvoed

philbucher commented on kratos/includes/initial_state.h:
```diff
@@ -0,0 +1,279 @@
+ * @details Storages the information regarding initial stresses/strains/F
+ * @author Alejandro Cornejo
+ */
+class KRATOS_API(KRATOS_CORE) InitialState
```
header source separation

philbucher commented on kratos/python_scripts/assign_initial_state_process.py:
```diff
@@ -0,0 +1,125 @@
+            msg = "SetInitialStateProcess: 'x', 'y' and 'z' variables not supported yet"
+            raise Exception(msg)
+        v.append(function)
+    return v
```
what does this do?
Cannot it be done in the same was as the other processes?

AlejandroCornejo replied:
@marandra

marandra replied:
> what does this do?

The `components_ to_string` function receives a vector as written by the user, and converts its components to strings, and in turn to functions. In this way, the user can pass a vector like [0.0, "1.2", "3.4 + t"] and it will be stored as [f(0), f(1.2), f(3.4+t)).
(I realize that it should be called something like "components_to _functions" and that its missing the docstrings.)

> Cannot it be done in the same was as the other processes?

I was looking for examples on how implement this functionality, I only found it used in the  `assign_scalar_variable_process.py` process, but it looked to me an ad-hoc implementation. If there is a recommended way to implement it, I will happily do it.

There are some more missing details regarding the process, like adapting it to 2D, and converting to functions the components of the matrix too. I'll push another commit with it.

philbucher commented on kratos/python_scripts/assign_initial_state_process.py:
```diff
@@ -0,0 +1,125 @@
+# Importing the Kratos Library
```
the name of this file is still wrong, should be `set_initial_state_process`

marandra replied:
I agree

AlejandroCornejo replied:
I'll do it

philbucher commented on kratos/python_scripts/assign_initial_state_process.py:
```diff
@@ -0,0 +1,125 @@
+    return v
+
+
+class SetInitialStateProcess(KratosMultiphysics.Process):
```
test is missing

loumalouomega replied:
Indeed

AlejandroCornejo replied:
@marandra can you do this? Otherwise I can find some time to do that

marandra replied:
sure

philbucher reviewed (commented):
for me only minor things are needed, see my comments
also @KlausBSautter must take a look to ensure it also works for him

loumalouomega commented on kratos/processes/set_initial_state_process.cpp:
```diff
@@ -0,0 +1,177 @@
+/***********************************************************************************/
+/***********************************************************************************/
+
+    template<std::size_t TDim>
```
Why this tabulations?

loumalouomega reviewed (changes requested):
Same comments of @philbucher + one comment

AlejandroCornejo committed:
parallel utilities

RiccardoRossi commented on applications/StructuralMechanicsApplication/custom_constitutive/elastic_isotropic_3d.cpp:
```diff
@@ -60,17 +60,19 @@ void  ElasticIsotropic3D::CalculateMaterialResponsePK2(ConstitutiveLaw::Paramete
         CalculateCauchyGreenStrain( rValues, r_strain_vector);
     }
 
+    this->AddInitialStrainVectorContribution( r_strain_vector, rValues );
```
do you need the "this"?

AlejandroCornejo replied:
nope, I can remove it -> BTW, this strain calculation has to be modified in the future. The correct way of doing it is to call the CalculateMaterialResponse deactivating th calculate stress and D.

RiccardoRossi commented on kratos/includes/initial_state.h:
```diff
@@ -0,0 +1,279 @@
+        ///@{
+        typedef std::size_t SizeType;
+
+        enum class InitialImposingType {StrainOnly = 0, StressOnly = 1, DeformationGradientOnly = 2, StrainAndStress = 3, DeformationGradientAndStress = 4};
```
this is an enum, so in kratos we normally use it all capitals, something like STRAIN_ONLY, STRESS_ONLY, DEFORMATION_GRADIENT_ONLY ...

AlejandroCornejo replied:
ok

marandra committed:
Python process updates:
-Rename process file
-Docstring
-Rename process' function
-Implement 2D vector to functions

RiccardoRossi commented on kratos/includes/initial_state.h:
```diff
@@ -0,0 +1,279 @@
+            const SizeType dimension = (voigt_size == 6) ? 3 : 2;
+
+            mInitialStrainVector.resize(voigt_size, false);
+            mInitialStressVector.resize(voigt_size, false);
```
I would argue that in this case you would best not resize what you don't need...

RiccardoRossi commented on kratos/includes/initial_state.h:
```diff
@@ -0,0 +1,279 @@
+            const SizeType voigt_size = rInitialStrainVector.size();
+            KRATOS_ERROR_IF(voigt_size <= 0) << "The imposed vector is null..." << std::endl;
+
+            mInitialStrainVector.resize(voigt_size, false);
```
you should set the initial size of the vector before resizing. If it has already the correct size it will be much faster

AlejandroCornejo replied:
You cannot know the initial size because when you create the object the dimnesion is unknown..

RiccardoRossi commented on kratos/includes/initial_state.h:
```diff
@@ -0,0 +1,279 @@
+            const SizeType voigt_size = rInitialStressVector.size();
+            KRATOS_ERROR_IF(voigt_size <= 0) << "The imposed vector is null..." << std::endl;
+
+            mInitialStressVector.resize(voigt_size, false);
```
same

RiccardoRossi commented on kratos/includes/initial_state.h:
```diff
@@ -0,0 +1,279 @@
+            const SizeType dimension = rInitialDeformationGradientMatrix.size1();
+            KRATOS_ERROR_IF(dimension <= 0) << "The imposed Matrix is null..." << std::endl;
+
+            mInitialDeformationGradientMatrix.resize(dimension, dimension, false);
```
same

RiccardoRossi commented on kratos/processes/set_initial_state_process.cpp:
```diff
@@ -0,0 +1,211 @@
+
+        if (InitialStateType == 0) {
+            noalias(mInitialStrain) = rInitialStateVector;
+            noalias(mInitialStress) = ZeroVector(voigt_size);
```
why do you set this? if you don't work with it you could leave it of size 0 and just check that at the moment of using it

AlejandroCornejo replied:
it can be done this way, yes. then there is an additional if in the constitutive law in order to check the size

RiccardoRossi reviewed (commented):
some comments.

AlejandroCornejo committed:
name changed of the pyton process

AlejandroCornejo committed:
separating in h and cpp the initial state

AlejandroCornejo committed:
removing this

AlejandroCornejo committed:
Changing the enum to uppercase

marandra committed:
Merge branch 'Structural/adding-initial-stress-strain-vectors' of github.com:KratosMultiphysics/Kratos into Structural/adding-initial-stress-strain-vectors

AlejandroCornejo committed:
custom strain size capability

philbucher commented on kratos/includes/initial_state.cpp:
```diff
@@ -0,0 +1,180 @@
+//    |  /           |
```
this file should be in `sources`

AlejandroCornejo committed:
moving to sources the cpp

marandra committed:
Missing include

marandra committed:
Merge branch 'Structural/adding-initial-stress-strain-vectors' of github.com:KratosMultiphysics/Kratos into Structural/adding-initial-stress-strain-vectors

philbucher reviewed (dismissed):
ok for me now

philbucher committed:
cleaning includes

maceligueta commented on kratos/processes/set_initial_state_process.cpp:
```diff
@@ -0,0 +1,210 @@
+        // for (int i = 0; i < static_cast<int>(mrModelPart.Elements().size()); i++) {
+        //     auto it_elem = it_elem_begin + i;
+
+        //     // If the values are set element-wise have priority
```
are these commented-out  lines necessary?

AlejandroCornejo replied:
yes, to see If you are **checking** ;) I'll remove them, THX

AlejandroCornejo committed:
removing commented code

philbucher reviewed (commented):
I think you are completely missing the serialization, or is there a specific reason you didn't add it?

AlejandroCornejo committed:
adding serializer

philbucher reviewed (commented):
cool
now you should also add the serialization to `ConstitutiveLaw`

AlejandroCornejo committed:
adding serializer

marandra committed:
Convert to functions deformation gradient components

AlejandroCornejo committed:
adding python test of the process

AlejandroCornejo committed:
adding python test

AlejandroCornejo committed:
minor

marandra committed:
Export initial state variables to python

loumalouomega committed:
Minor to "block merge"

philbucher commented on kratos/sources/initial_state.cpp:
```diff
@@ -0,0 +1,180 @@
+
+    // Selective constructor for vectors
+    InitialState::InitialState(const Vector& rImposingEntity,
+                    const int InitialImposingType = static_cast<int>(InitialImposingType::STRAIN_ONLY))
```
while looking at the CentOS error I noticed all the casts here
@AlejandroCornejo those should not be there, you can directly work with the `enum`
can you please fix this in a follow up PR?
